### ISSUE TITLE
Partitioning of the FK tables?

### ISSUE THREAD

**Issue by yazun** [moat/range_partitioning#1]
This is more feature request/question than issue:
what happens with the tables that reference partitioned table via fk which is a partition key: will they be partitioned too? what happens to fk's in general in such a situation?

**Comment by coreyatmoat** [comment]
Foreign Key references which point to a partitioned table will not work: the table that previously housed the data is now empty, and the child tables hold all of the data. 

This is true of any partitioning scheme currently available in vanilla postgresql. 

As to what specifically would happen:
If the table already contained data referenced by another table, then the call to create_parent() would fail when the rows in the parent table are moved to the first child partition via delete/returning/insert. 
If the table lacked any data referenced by another table, then the create_parent() and subsequent create_partition() calls would succeed. However, no rows could be added to the referencing table which had a non-null value in the referencing column.

**Comment by yazun** [comment]
Thanks for the clarification. It would be good (although clearly non-trivial) to cascade partitioning to all referencing tables: obviously FKs would not make sense otherwise. 
This also brings the question of nested partitioning - some partitioning scheme could be independently defined on the referencing tables, but the same tables would have to be partitioned on the referred table partitioning scheme first. Do you think this would be doable?

**Comment by coreyatmoat** [comment]
If table A references table B on column C, and we create_parent('A','C'), that's no problem at all. The table A_p0 will have all of the data previously contained in A, and A_p0 will inherit the foreign key constraint from A.

The problem lies in going the other direction: partitioning B is problematic and basically not do-able while still enforcing generic foreign keys.

It would be possible to manually create foreign keys from A_p0 to B_p0, A_p1 to B_p1, etc. These would be constraints that do not exist in the parent tables A and B, and the synchronization of ranges would be up to the user, and probably easiest if the partitioning scheme were static. Still, if you made the changes in a transaction and either dropped or deferred the constraints until commit-time, it would be possible and would look something like this (below is pseudo-code, not tested syntax):

```sql
begin;
select range_partitioning.create_parent('a','c')
select range_partitioning.create_parent('b','c');
select range_partitioning.create_partition('a','(,10000)');
select range_partitioning.create_partition('b','(,10000)');
alter table a_p0 add constraint a_p0_fk foreign key references b_p0(c);
alter table a_p1 add constraint a_p0_fk foreign key references b_p1(c);
end;
```

Again, do-able, but not seamless.

**Comment by yazun** [comment]
Yes, exactly: why we could not traverse the FK relationships and partition all children on the path and automatically create referencing partitions?

**Comment by coreyatmoat** [comment]
It does appear possible. However, I strongly doubt that the query planner would recognize the relationship and push-down the join of the two parent tables to the corresponding partitions, so it would only serve data-integrity purposes.

**Comment by yazun** [comment]
Judging by experience it should, if the condition matches the partition exclusion clause, which should be good enough in most of the cases.
I can try to post an example.

**Comment by yazun** [comment]
So what will happen with Fks on existing tables when range_partitioning.create_partition is run on B?

**Comment by chuinker** [comment]
Any row in B that is referenced in A must remain in B itself, not a partition. So if there were no rows in A referencing B (i.e. all FK values were null) the operation would succeed. If any one value wasn't null, the operation would fail.
